paper_framework_sources/p1897_initial_executor_algorithms_for_cpp23.md

```diff
@@ -694,21 +748,123 @@ float r = sync_wait(
 
 ### Wording
 The name `execution::handle_error` denotes a customization point object.
-The expression `execution::handle_error(S, F)` for some subexpressions S and F is expression-equivalent to:
+For some subexpressions `s` and `f`, let `S` be a type such that `decltype((s))` is `S` and `decltype((f))` is `F`.
+The expression `execution::handle_error(s, f)` is expression-equivalent to:
 
- * S.handle_error(F), if that expression is valid.
- * Otherwise, `handle_error(S, F)`, if that expression is valid with overload resolution performed in a context that includes the declaration
+ * s.handle_error(f), if that expression is valid, if `s` satisfies `sender` and `f` satisfies `invocable`.
+ * Otherwise, `handle_error(s, f)`, if that expression is valid,, if `s` satisfies `sender` and `f` satisfies `invocable` with overload resolution performed in a context that includes the declaration
 ```
-        template<class S, class F>
-          void handle_error(S, F) = delete;
+    void handle_error() = delete;
 ```
   and that does not include a declaration of `execution::handle_error`.
+ * Otherwise constructs a receiver, `r` and passes that receiver to `execution::connect(S, r)` returning an `operation_state` `os` such that
+   When some `output_receiver` has been passed to `connect` on the returned `sender` returning some `operation_state` `os2`:
 
- * Otherwise constructs a receiver, `r` over an implementation-defined synchronization primitive and passes that receiver to `execution::submit(S, r)`.
+   * If `set_value(r, ts...)` is called, passes `ts...` to `set_value(output_receiver, ts...)`.
+   * If `set_error(r, e)` is called, calls `std::invoke(f, ts...)` to return some `invoke_result`, and calls `execution::start(execution::connect(invoke_result, output_receiver))`.
+   * If `f` throws, catches the exception and passes it to `set_error(output_receiver, e)`.
+   * If `set_done(r)` is called, calls `set_done(output_receiver)`.
+
+   When `start` is called on `os2`, call `execution::start(os)`.
+
+ * Otherwise the expression `execution::handle_error(s, f)` is ill-formed.
+
+
+## execution::share
+
+### Overview
+`share` is a sender adapter that takes a `sender`, eagerly submits it and returns a `sender` that propagates the value by reference and can be used as an l-value.
+
+Signature:
+```cpp
+template <execution::sender S>
+see-below share(S s);
+```
+
+*[ Example:*
+```cpp
+auto s1 = just(3) | share();
+auto s2 = s1 | transform([](const int& v){return v+1;}))
+auto s3 = s1 | transform([](const int& v){return v+2;}))
+int r = sync_wait(
+  transform(
+    when_all(s2, s3),
+    [](int a, int b){return a+b;}));
+// r==9
+```
```

Comment:
in order for this to behave as expected share() must store the value (this is what replay() and replay_subject do - share() normally does not).

The issue is that share() is ref counted. so when_all(s2, s3) will start s2, which will start share() with one ref which will start just() and perform the transform(), then complete just() and return to share() which will refcount to 0 and return to when_all() which will start s3. s3 will start share() with one ref which will start just() and perform the transform(), then complete just() and return to share() which will refcount to 0 and return to when_all() which will complete and perform the final transform.

This is another reason that rx adds a separate connect task to publish and then has a separate ref_count() algo and those are composed into share(). 

```cpp
auto s1 = just(3) | publish();
auto s2 = s1 | transform([](const int& v){return v+1;}))
auto s3 = s1 | transform([](const int& v){return v+2;}))
int r = sync_wait(
  transform(
    when_all(s2, s3, s1.run()),
    [](int a, int b){return a+b;}));
// r==9
```

now the just is not started until after both s2 and s3 are listening

Reply:
Yes, the actual expected semantics of share was an interesting question.

So in this case this is inefficient because of the refcounting, but it's only a trivial example. More realistically maybe separate sync_waits of s2 and s2 would be more appropriate as the use case would be sending the result elsewhere. So share is taking a potentially r-value connectable sender and making it copyable. In that example the publish approach is less obviously useful because it requires coordination with all the downstream work. 

My interpretation was that share produced something like a shared_future - started and reusable both before and after it completes.

Reply:
Updated as ensure_started and without reference counting as discussed offline.